### DISCUSSION TITLE
Error on Moving from Sequential to LocalExecutor

### DISCUSSION THREAD

**Discussion by sheharbano96** [apache/airflow#44243] [Q&A]
I just updated my airflow setup to use LocalExecutor but it started giving this error: 
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
AttributeError: 'QueuedLocalWorker' object has no attribute 'wrapper'
```
It works fine with Sequential Executor but keeps queuing tasks with this error on the scheduler.

**Comment by Shubhisnottaken** [comment]
have you resolved this?

**Reply by sheharbano96** [reply]
No.

**Reply by Shubhisnottaken** [reply]
use Python 3.10.12 and AIRFLOW_VERSION=2.10.2.

**Reply by sheharbano96** [reply]
My python is 3.10.6 and airflow is 2.10.3

**Reply by keichi** [reply]
I'm also facing this issue with Python 3.10.16 and Airflow 2.10.2. Are there any workarounds?

**Reply by Shubhisnottaken** [reply]
Python 3.10.12  and airflow 2.10.2 working for me.

**Reply by potiuk** [reply]
I can think of a few things that could cause similar issues

1. Make sure you are installing Airlfow with the right constraints https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html
2. Possibly you are using some limited Python interpreter delivered by the distribution you are using (your path indicate MacOS). Make sure you install Python interpreter and create a virtual environment where you install Airlfow (you can use `uv` to install Python and create virtualenvs where you install Airflow). The Python distributions delivered by OS are often limited and trimmed down and have some "internals" changed to use Python for some internal scripts only - so if you want to run a complex Python application, make sure to install a "vanilla" interpreter - not the one modified by the OS producer
3. If you are using different OS than Linux, there might be various problems - Windows is not supported at all, WSL2 can be used for debugging and running a "quick start" - similarly MacOS has problems with forking in Python that seriously limits stability and usability of Airflow for "production" use cases. Practically only Linux distros are recommended for any production grade running of Airflow.

**Reply by keichi** [reply]
Thanks for you advice, I am using macOS so 3. might be the issue. 1. and 2. should be fine since I installed Python via Homebrew and Airflow via PyPI using the constraints file. I'll try on a Linux system.

**Reply by potiuk** [reply]
> Thanks for you advice, I am using macOS so 3. might be the issue. 1. and 2. should be fine since I installed Python via Homebrew and Airflow via PyPI using the constraints file. I'll try on a Linux system.

Looking at the paths - you ARE using the built-in Python

`/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/spawn.py"`

This - very likely - is the system Python.

I guess even if you have installed Airflow with brew it might not be used by default. On top of it you are not using virtualenv - which is generaly a bad idea, - installng application in "python" without creating venv is generally very strongly discourage by the `pip` team and anyone from packaging team. It causes a LOT of problems when you mix system dependencies and `pip` installed packages. Latest version of `pip` even does not allow you to do it and you need to specigy `--allow-breaking-system-packages` or smth like that flag when you try to do it.

I strongly recommend using `uv` including `uv python`  and looking at setting up venv with it. It's lightning fast (comparing to brew managed python) and it has nice workflow with venv management

**Reply by keichi** [reply]
I think you are confusing me with @sheharbano96

**Reply by potiuk** [reply]
Ah, Might be. That's always a risk when you comment on other's thread and continue conversation with your case. Sometimes it's worth to start new discussion with "relate to #` to avoid it. When you have 10-15 messages in a discussion and as maintainer you handle 100s of them a day, clear separation helps, so that you do not have to re read the whole thread every time you come back to it.

Just for the future :)

**Reply by sheharbano96** [reply]
@potiuk Yes, I am using macOS but I have installed Airflow in the python's virtual environment. I'll try updating the interpreter and reinstalling airflow with constraints and see if the issue persists.  
I will be running it on Linux for production but testing it on my local which is macOS. The SequentialExecutor doesn't work for sending slack updates so I was trying to update it to LocalExecutor to test slack updates.
Thank you for your response.

**Reply by hshabi** [reply]
I'm having the same issue, with macOS.
I installed it on the virtual environments and tried Python 3.11.8 / 3.10.12, but it didn't work.
SequentialExecutor worked fine.
Before the upgrade to Airflow 2.10.2, I used Airflow 2.6.3 with Python 3.8.18, and the localexecutor worked fine without errors.
Any solution on how to use localexecutor?

**Reply by pbjc** [reply]
I think the problem is that the `span` decorator does not use `functools.wraps`, making the decorated methods unable to be serialized by pickle. Removing `@span` from `LocalWorkerBase` and its child classes fixed the issue for me ([ex](https://github.com/apache/airflow/blob/c99887ec11ce3e1a43f2794fcf36d27555140f00/airflow/executors/local_executor.py#L82)). I'm currently using a modified `LocalExecutor` with `AIRFLOW__CORE__EXECUTOR` to get around this issue locally.

**Reply by jhb115** [reply]
I am facing the exact same issue. I hope this will be fixed soon.

**Reply by potiuk** [reply]
Can someone here submit a fix for that please after testing it locally? That would definitely help in fixing it faster.

**Reply by lok419** [reply]
same issue here, looks like MacOS only

**Reply by tony-brookes** [reply]
> I think the problem is that the `span` decorator does not use `functools.wraps`, making the decorated methods unable to be serialized by pickle. Removing `@span` from `LocalWorkerBase` and its child classes fixed the issue for me ([ex](https://github.com/apache/airflow/blob/c99887ec11ce3e1a43f2794fcf36d27555140f00/airflow/executors/local_executor.py#L82)). I'm currently using a modified `LocalExecutor` with `AIRFLOW__CORE__EXECUTOR` to get around this issue locally.

Do you have any details on how you did this?  Did you just copy the local_executor.py file somewhere locally and then update your configuration to point to it?  The configuration seems to use an unqualified class name by default so I'm not sure what the correct configuration would be to set it to my custom class.

**Reply by pbjc** [reply]
@tony-brookes The core executors that ship with Airflow have aliases which are defined [here](https://github.com/apache/airflow/blob/2.10.5/airflow/executors/executor_loader.py#L63).

Otherwise the supported ways Airflow can find your executor depends on the version: https://github.com/apache/airflow/blob/2.10.5/airflow/executors/executor_loader.py#L218

I just use the import path approach

**Reply by Mahesh-01010011** [reply]
@tony-brookes, were you able to figure out the code needed based on what pbjc said? It was vague to me. It would have been more useful if he had shared the code snippet that worked for him, but as a beggar, I can't be chooser.

Based on his other comment about `wraps`, I opened the `/path/to/venv-3.11-airflow-2.10/lib/python3.11/site-packages/airflow/traces/tracer.py` file and added the `@functools.wraps` decorator inside `span` function like this:
```python
def span(func):
    """Decorate a function with span."""

    import functools
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        ...
```

It's not an elegant solution, but it works for me. I'm not sure if doing another `pip install` will override this, but that's ok for me. My Mac is just a dev environment for me. I'll add these lines again to get it to work.

**Comment by worst-username-ever** [comment]
I had the same issue and tried commenting out some Darwin-specific logic in the multiprocessing module in cpython 3.12, as it seemed to be deprecated by now. Specifically in those files:
https://github.com/python/cpython/blob/3.12/Lib/multiprocessing/context.py
https://github.com/python/cpython/blob/3.12/Lib/multiprocessing/reduction.py

This seemed to fix the issue for me. However, please be aware that I have absolutely no idea what this code does let alone which bugs it was trying to fix. I have almost zero Python experience so I feel unsure about submitting a PR for a change which may cause more problems than it solves.

**Reply by Samych02** [reply]
That's the only solution that seemed to work, thanks a lot!
